src/Features/Core/Portable/ExtractInterface/AbstractExtractInterfaceService.cs

```diff
@@ -177,20 +214,65 @@ public ExtractInterfaceResult ExtractInterfaceFromAnalyzedType(ExtractInterfaceT
+                extractInterfaceOptions.IncludedMembers,
+                solution,
+                refactoringResult.TypeNode,
+                cancellationToken);
+
+            var document = annotatedSolution.GetDocument(refactoringResult.DocumentToExtractFrom.Id);
+            var originalRoot = document.GetSyntaxRootSynchronously(cancellationToken);
+            var typeDeclaration = originalRoot.GetAnnotatedNodes(typeNodeSyntaxAnnotation).Single();
+
+            var trackedDocument = document.WithSyntaxRoot(originalRoot.TrackNodes(typeDeclaration));
+
+            var currentRoot = trackedDocument.GetSyntaxRootSynchronously(cancellationToken);
+            var editor = new SyntaxEditor(currentRoot, annotatedSolution.Workspace);
+
+            // Generate the interface syntax node, which will be inserted above the type it's extracted from
+            var codeGenService = trackedDocument.GetLanguageService<ICodeGenerationService>();
+            var interfaceNode = codeGenService.CreateNamedTypeDeclaration(extractedInterfaceSymbol).WithAdditionalAnnotations(SimplificationHelpers.SimplifyModuleNameAnnotation);
+
+            typeDeclaration = currentRoot.GetCurrentNode(typeDeclaration);
+            editor.InsertBefore(typeDeclaration, interfaceNode);
+
+            var unformattedSolution = document.WithSyntaxRoot(editor.GetChangedRoot()).Project.Solution;
+
+            // After the interface is inserted, update the original type to show it implements the new interface
+            var completedSolution = GetSolutionWithOriginalTypeUpdated(
+                unformattedSolution, documentIds,
+                refactoringResult.DocumentToExtractFrom.Id, typeNodeSyntaxAnnotation,
+                refactoringResult.TypeToExtractFrom, extractedInterfaceSymbol,
+                extractInterfaceOptions.IncludedMembers, symbolToDeclarationAnnotationMap, cancellationToken);
+
+            return new ExtractInterfaceResult(
+                succeeded: true,
+                updatedSolution: completedSolution,
+                navigationDocumentId: refactoringResult.DocumentToExtractFrom.Id);
+        }
+
+        private 
+            (Dictionary<ISymbol, SyntaxAnnotation> symbolToDeclarationAnnotationMap, 
+            Solution annotatedSolution, 
+            List<DocumentId> documentIds, 
+            SyntaxAnnotation typeNodeAnnotation) 
+            CreateSymbolToDeclarationAnnotationMap(
+                IEnumerable<ISymbol> includedMembers,
+                Solution solution,
+                SyntaxNode typeNode,
+                CancellationToken cancellationToken)
         {
             var symbolToDeclarationAnnotationMap = new Dictionary<ISymbol, SyntaxAnnotation>();
             var currentRoots = new Dictionary<SyntaxTree, SyntaxNode>();
-            documentIds = new List<DocumentId>();
+            var documentIds = new List<DocumentId>();
 
             var typeNodeRoot = typeNode.SyntaxTree.GetRoot(CancellationToken.None);
```

Comment:
asynchronous and synchronous APIs are all mixed used here. is it like depending on the implementation detail that typenode will holding onto all its spine so that root will be in memory?